[pkg/stanza] Add cache for `container` operator's k8s attributes extraction

Nitesh-vaidyanath opened the pull request:
<!--Ex. Fixing a bug - Describe the bug and how this fixes the issue.
Ex. Adding a feature - Explain what this achieves.-->
#### Description

<!-- Issue number (e.g. #1234) or full URL to issue, if applicable. -->
#### Link to tracking issue

[#44486](https://github.com/open-telemetry/opentelemetry-collector-contrib/issues/44486)

<!--Describe what testing was performed and which tests were added.-->
#### Testing
```
go test -v .
=== RUN   TestNewMemoryCache
=== RUN   TestNewMemoryCache/size-50
=== RUN   TestNewMemoryCache/size-1024
--- PASS: TestNewMemoryCache (0.00s)
    --- PASS: TestNewMemoryCache/size-50 (0.00s)
    --- PASS: TestNewMemoryCache/size-1024 (0.00s)
=== RUN   TestMemoryCacheGetAndAdd
=== RUN   TestMemoryCacheGetAndAdd/basic
--- PASS: TestMemoryCacheGetAndAdd (0.00s)
    --- PASS: TestMemoryCacheGetAndAdd/basic (0.00s)
=== RUN   TestMemoryCacheFIFOEviction
--- PASS: TestMemoryCacheFIFOEviction (0.00s)
=== RUN   TestMemoryCacheCopy
--- PASS: TestMemoryCacheCopy (0.00s)
=== RUN   TestMemoryCacheMaxSize
--- PASS: TestMemoryCacheMaxSize (0.00s)
=== RUN   TestNewStartedAtomicLimiter
=== RUN   TestNewStartedAtomicLimiter/default
=== RUN   TestNewStartedAtomicLimiter/max
=== RUN   TestNewStartedAtomicLimiter/interval
--- PASS: TestNewStartedAtomicLimiter (0.00s)
    --- PASS: TestNewStartedAtomicLimiter/default (0.00s)
    --- PASS: TestNewStartedAtomicLimiter/max (0.00s)
    --- PASS: TestNewStartedAtomicLimiter/interval (0.00s)
=== RUN   TestLimiter
--- PASS: TestLimiter (0.00s)
=== RUN   TestThrottledLimiter
--- PASS: TestThrottledLimiter (2.00s)
=== RUN   TestThrottledCache
--- PASS: TestThrottledCache (0.00s)
=== RUN   TestMemoryCacheConcurrentAccess
--- PASS: TestMemoryCacheConcurrentAccess (0.00s)
=== RUN   TestConfig
=== RUN   TestConfig/default
=== RUN   TestConfig/parse_from_simple
=== RUN   TestConfig/parse_to_simple
=== RUN   TestConfig/on_error_drop
=== RUN   TestConfig/severity
=== RUN   TestConfig/format
=== RUN   TestConfig/add_metadata_from_file_path
=== RUN   TestConfig/max_log_size
=== RUN   TestConfig/parse_to_attributes
=== RUN   TestConfig/parse_to_body
=== RUN   TestConfig/parse_to_resource
--- PASS: TestConfig (0.00s)
    --- PASS: TestConfig/default (0.00s)
    --- PASS: TestConfig/parse_from_simple (0.00s)
    --- PASS: TestConfig/parse_to_simple (0.00s)
    --- PASS: TestConfig/on_error_drop (0.00s)
    --- PASS: TestConfig/severity (0.00s)
    --- PASS: TestConfig/format (0.00s)
    --- PASS: TestConfig/add_metadata_from_file_path (0.00s)
    --- PASS: TestConfig/max_log_size (0.00s)
    --- PASS: TestConfig/parse_to_attributes (0.00s)
    --- PASS: TestConfig/parse_to_body (0.00s)
    --- PASS: TestConfig/parse_to_resource (0.00s)
=== RUN   TestConfigBuild
--- PASS: TestConfigBuild (0.00s)
=== RUN   TestConfigBuildFailure
--- PASS: TestConfigBuildFailure (0.00s)
=== RUN   TestConfigBuildFormatError
--- PASS: TestConfigBuildFormatError (0.00s)
=== RUN   TestDockerParserInvalidType
--- PASS: TestDockerParserInvalidType (0.00s)
=== RUN   TestCrioParserInvalidType
--- PASS: TestCrioParserInvalidType (0.00s)
=== RUN   TestContainerdParserInvalidType
--- PASS: TestContainerdParserInvalidType (0.00s)
=== RUN   TestFormatDetectionFailure
--- PASS: TestFormatDetectionFailure (0.00s)
=== RUN   TestInternalRecombineCfg
--- PASS: TestInternalRecombineCfg (0.00s)
=== RUN   TestProcess
=== RUN   TestProcess/Success
=== RUN   TestProcess/Success/docker
=== RUN   TestProcess/Success/docker_with_auto_detection
=== RUN   TestProcess/Success/docker_with_auto_detection_and_metadata_from_file_path
=== RUN   TestProcess/Failure
=== RUN   TestProcess/Failure/docker_with_add_metadata_from_filepath_but_not_included
--- PASS: TestProcess (0.00s)
    --- PASS: TestProcess/Success (0.00s)
        --- PASS: TestProcess/Success/docker (0.00s)
        --- PASS: TestProcess/Success/docker_with_auto_detection (0.00s)
        --- PASS: TestProcess/Success/docker_with_auto_detection_and_metadata_from_file_path (0.00s)
    --- PASS: TestProcess/Failure (0.00s)
        --- PASS: TestProcess/Failure/docker_with_add_metadata_from_filepath_but_not_included (0.00s)
=== RUN   TestRecombineProcess
=== RUN   TestRecombineProcess/crio_standalone_with_auto_detection_and_metadata_from_file_path
=== RUN   TestRecombineProcess/crio_standalone_with_auto_detection_and_metadata_from_rotated_file_path
=== RUN   TestRecombineProcess/containerd_standalone_with_auto_detection_and_metadata_from_file_path
=== RUN   TestRecombineProcess/containerd_standalone_with_auto_detection_and_metadata_from_rotated_file_path
=== RUN   TestRecombineProcess/crio_multiple_with_auto_detection_and_metadata_from_file_path
=== RUN   TestRecombineProcess/containerd_multiple_with_auto_detection_and_metadata_from_file_path
=== RUN   TestRecombineProcess/containerd_multiple_with_auto_detection_and_metadata_from_file_path_windows
--- PASS: TestRecombineProcess (0.71s)
    --- PASS: TestRecombineProcess/crio_standalone_with_auto_detection_and_metadata_from_file_path (0.10s)
    --- PASS: TestRecombineProcess/crio_standalone_with_auto_detection_and_metadata_from_rotated_file_path (0.10s)
    --- PASS: TestRecombineProcess/containerd_standalone_with_auto_detection_and_metadata_from_file_path (0.10s)
    --- PASS: TestRecombineProcess/containerd_standalone_with_auto_detection_and_metadata_from_rotated_file_path (0.10s)
    --- PASS: TestRecombineProcess/crio_multiple_with_auto_detection_and_metadata_from_file_path (0.10s)
    --- PASS: TestRecombineProcess/containerd_multiple_with_auto_detection_and_metadata_from_file_path (0.10s)
    --- PASS: TestRecombineProcess/containerd_multiple_with_auto_detection_and_metadata_from_file_path_windows (0.10s)
=== RUN   TestProcessWithDockerTime
=== RUN   TestProcessWithDockerTime/docker
--- PASS: TestProcessWithDockerTime (0.00s)
    --- PASS: TestProcessWithDockerTime/docker (0.00s)
=== RUN   TestCRIRecombineProcessWithFailedDownstreamOperator
=== RUN   TestCRIRecombineProcessWithFailedDownstreamOperator/crio_multiple
=== RUN   TestCRIRecombineProcessWithFailedDownstreamOperator/containerd_multiple
--- PASS: TestCRIRecombineProcessWithFailedDownstreamOperator (0.20s)
    --- PASS: TestCRIRecombineProcessWithFailedDownstreamOperator/crio_multiple (0.10s)
    --- PASS: TestCRIRecombineProcessWithFailedDownstreamOperator/containerd_multiple (0.10s)
=== RUN   TestPathCacheIntegration
=== RUN   TestPathCacheIntegration/CacheEnabled
=== RUN   TestPathCacheIntegration/CacheDisabled
=== RUN   TestPathCacheIntegration/CacheReusesParsedPaths
=== RUN   TestPathCacheIntegration/CacheHandlesDifferentPaths
=== RUN   TestPathCacheIntegration/CacheFIFOEviction
=== RUN   TestPathCacheIntegration/CacheStop
--- PASS: TestPathCacheIntegration (0.01s)
    --- PASS: TestPathCacheIntegration/CacheEnabled (0.00s)
    --- PASS: TestPathCacheIntegration/CacheDisabled (0.00s)
    --- PASS: TestPathCacheIntegration/CacheReusesParsedPaths (0.00s)
    --- PASS: TestPathCacheIntegration/CacheHandlesDifferentPaths (0.00s)
    --- PASS: TestPathCacheIntegration/CacheFIFOEviction (0.00s)
    --- PASS: TestPathCacheIntegration/CacheStop (0.00s)
PASS
ok      github.com/open-telemetry/opentelemetry-collector-contrib/pkg/stanza/operator/parser/container    3.335s

```
<!--Describe the documentation added.-->
#### Documentation

<!--Please delete paragraphs that you did not use before submitting.-->

Nitesh-vaidyanath committed:
added cache for container operator k8s attribtues extraction

Nitesh-vaidyanath commented on pkg/stanza/operator/parser/container/package_test.go:
```diff
@@ -10,5 +10,10 @@ import (
+		// Ignore goroutines from the atomicLimiter background ticker
+		// These goroutines are properly stopped when cache.stop() is called,
+		// but may still be exiting when goleak checks
+		goleak.IgnoreTopFunction("github.com/open-telemetry/opentelemetry-collector-contrib/pkg/stanza/operator/parser/container.(*atomicLimiter).init.func1.1"),
```
@atoulme Yeah for cache.stop I have ignored because goroutine exits when done is closed without that it would return goleak so had to ignore it.

atoulme commented on pkg/stanza/operator/parser/container/cache.go:
```diff
@@ -0,0 +1,220 @@
+
+// newMemoryCache takes a cache size and a limiter interval and
+// returns a new memory backed cache
+func newMemoryCache(maxSize uint16, interval uint64) *memoryCache {
```
is there no lru implementation lib you could use instead of creating your own?

Nitesh-vaidyanath replied:
> Understood. Since this is a copy of the regex operator cache, any way to reuse that code and avoid duplication here?

Yeah we should be, let me see what I can do here

Nitesh-vaidyanath commented on pkg/stanza/operator/parser/container/cache.go:
```diff
@@ -0,0 +1,220 @@
+var _ limiter = &atomicLimiter{count: &atomic.Uint64{}}
+
+// init initializes the limiter
+func (l *atomicLimiter) init() {
```
No same copied same code from pkg/stanza/operator/parser/regex/cache.go.

Nitesh-vaidyanath committed:
moved pathCache.stop() before that guard so it always runs

Nitesh-vaidyanath committed:
moved cache under helper for both regex and container and added seperate methods for sync map and map cache

Nitesh-vaidyanath committed:
added change log

Nitesh-vaidyanath committed:
modified component to filelog

Nitesh-vaidyanath committed:
Merge branch 'main' into main

Nitesh-vaidyanath committed:
receiver/filelog: add fast‑path parser and shared cache to eliminate regex overhead

Nitesh-vaidyanath commented:
Tested docker build versions are available [here](https://hub.docker.com/repository/docker/niteshbv11/otelcontribcol/general).


**Baseline (v0.138)**

- Implementation: The container log parser relied heavily on regular expressions to extract Kubernetes metadata from file paths.  Every log line triggered a new regex match and there was no caching.
- CPU profile: About half of the CPU time was spent in helper.(*TransformerOperator).ProcessBatchWith, while regex backtracking functions consumed roughly one quarter to one third of CPU time.  This made metadata extraction a significant but secondary hotspot.
- Memory profile: The system allocated more than 365 billion objects during the sample period.  The extractk8sMetaFromFilePath function accounted for roughly 12 % of total allocations.


**v1 – Introduce caching of regex results**

- Changes: Introduced a cache (a map guarded by a read/write mutex) to store parsed Kubernetes attributes and refactored extractk8sMetaFromFilePath to factor out metadata setting into a helper (setK8sMetadataFromParsedValues).
- Test coverage: Added cache_test.go to validate cache initialization, FIFO eviction and thread‑safety.  Added TestPathCacheIntegration to ensure the parser reuses cached results when the same container path is encountered.
- Performance impact: CPU usage fell by about 25 %, with regex backtracking dropping from roughly 25 % to 18 % of CPU.  Total allocations decreased by ~22 %, from 365 billion to 285 billion objects, because repeated regex matches were avoided and metadata setting was consolidated.

<img width="1554" height="2552" alt="flamegraph_baseline_2025-11-26_1415-to-2025-11-26_1430_comparison_2025-11-26_1515-to-2025-11-26_1530_cpu" src="https://github.com/user-attachments/assets/0abcb4ca-17b0-4605-bd4e-5f70196d681b" />

<img width="1554" height="2156" alt="flamegraph_baseline_2025-11-26_1415-to-2025-11-26_1430_comparison_2025-11-26_1515-to-2025-11-26_1530_memory" src="https://github.com/user-attachments/assets/a2180938-1053-4b7d-bfc8-f7e158705d38" />

**v2 – Fast path: remove regex as default**

- Changes: Implemented a fast‑path parser using string splitting to extract namespace, pod UID, pod name and container name.  A temporary map of parsed values is built on each call, regex is now a fallback for unexpected formats.
- Test coverage: Updated parser tests to verify that the fast path extracts metadata correctly, while existing CRI and Windows path tests continued to pass.
- Performance impact: Eliminating regex from the primary path reduced CPU time by another ~19 % relative to v1 and removed regex functions from the hot path.  However, because a new map was created every time, total allocations increased by ~60 %.  In other words, CPU improved but memory usage grew.

<img width="1554" height="2596" alt="flamegraph_baseline_2025-11-26_1515-to-2025-11-26_1530_comparison_2025-11-26_1538-to-2025-11-26_1553_cpu" src="https://github.com/user-attachments/assets/489f77ee-a7d9-42ad-b699-6cb40892571b" />
<img width="1554" height="2156" alt="flamegraph_baseline_2025-11-26_1515-to-2025-11-26_1530_comparison_2025-11-26_1538-to-2025-11-26_1553_memory" src="https://github.com/user-attachments/assets/3805138a-f1ef-4d3e-8c20-42d60d6c9585" />


**v3 – Caching with concurrent map and object pooling**

Changes: Replaced per‑call map creation with a concurrent cache based on sync.Map and added object pools (sync.Pool) to reuse temporary structures.  The fast path remains primary, with regex as a fallback only.
Performance impact: CPU dropped by an additional ~4 % compared with v2, achieving roughly a 30 % reduction from the baseline.  Memory allocations decreased by ~10 % compared with v2 and ~9 % relative to v0.138.  Once regex overhead was removed, compression and file polling emerged as the main CPU consumers.

<img width="1554" height="2596" alt="flamegraph_baseline_2025-11-26_1538-to-2025-11-26_1553_comparison_2025-11-26_1615-to-2025-11-26_1630_cpu" src="https://github.com/user-attachments/assets/79b9eb50-b349-4e8b-a6b9-c635938bc8b6" />

<img width="1554" height="2156" alt="flamegraph_baseline_2025-11-26_1538-to-2025-11-26_1553_comparison_2025-11-26_1615-to-2025-11-26_1630_memory" src="https://github.com/user-attachments/assets/c7ee223f-b2c0-4738-a6d8-cec70e95d3a9" />

**v0.138 vs v3**

This direct comparison illustrates how much the parser has improved:

- CPU: Total sample time fell from 8 598.53 s in the baseline to 5 978.09 s with v3, confirming a ~30 % reduction.  Regex functions that previously consumed over 32 % of CPU disappeared entirely.
- Memory: Total allocation objects declined from 96.99 billion to 87.83 billion, ~9.4 % decrease.  While extractk8sMetaFromFilePath now accounts for 16.35 % of allocations (up from 12.16 %), its absolute allocation count still decreased because other parts of the system allocate far less. 

<img width="1554" height="2552" alt="flamegraph_baseline_2025-11-26_1415-to-2025-11-26_1430_comparison_2025-11-26_1615-to-2025-11-26_1630_cpu" src="https://github.com/user-attachments/assets/9b4436eb-7886-4a5e-9771-8ce69c18e54c" />
<img width="1554" height="2156" alt="flamegraph_baseline_2025-11-26_1415-to-2025-11-26_1430_comparison_2025-11-26_1615-to-2025-11-26_1630_memory" src="https://github.com/user-attachments/assets/b6afb931-cd57-469d-9e07-ec1beab278f4" />


**Summary:**

1. v1 introduced a simple cache, cutting CPU by ~25 % and reducing allocations by ~22 %.
2. v2 replaced regex with a string‑split fast path, shaving off another ~19 % of CPU but increasing allocations due to per‑call map creation.
3. v3 added a concurrent cache and object pooling, achieving roughly a 30 % CPU reduction in total while bringing memory allocations down below the baseline despite the added caching.

These optimisations, verified by new tests for caching and concurrency, ensure that CRI, generic container and Windows paths are parsed correctly and far more efficiently.


**Test Case:**
```
container % go test ./... -v
=== RUN   TestConfig
=== RUN   TestConfig/default
=== RUN   TestConfig/parse_from_simple
=== RUN   TestConfig/parse_to_simple
=== RUN   TestConfig/on_error_drop
=== RUN   TestConfig/severity
=== RUN   TestConfig/format
=== RUN   TestConfig/add_metadata_from_file_path
=== RUN   TestConfig/max_log_size
=== RUN   TestConfig/parse_to_attributes
=== RUN   TestConfig/parse_to_body
=== RUN   TestConfig/parse_to_resource
--- PASS: TestConfig (0.00s)
    --- PASS: TestConfig/default (0.00s)
    --- PASS: TestConfig/parse_from_simple (0.00s)
    --- PASS: TestConfig/parse_to_simple (0.00s)
    --- PASS: TestConfig/on_error_drop (0.00s)
    --- PASS: TestConfig/severity (0.00s)
    --- PASS: TestConfig/format (0.00s)
    --- PASS: TestConfig/add_metadata_from_file_path (0.00s)
    --- PASS: TestConfig/max_log_size (0.00s)
    --- PASS: TestConfig/parse_to_attributes (0.00s)
    --- PASS: TestConfig/parse_to_body (0.00s)
    --- PASS: TestConfig/parse_to_resource (0.00s)
=== RUN   TestConfigBuild
--- PASS: TestConfigBuild (0.00s)
=== RUN   TestConfigBuildFailure
--- PASS: TestConfigBuildFailure (0.00s)
=== RUN   TestConfigBuildFormatError
--- PASS: TestConfigBuildFormatError (0.00s)
=== RUN   TestDockerParserInvalidType
--- PASS: TestDockerParserInvalidType (0.00s)
=== RUN   TestCrioParserInvalidType
--- PASS: TestCrioParserInvalidType (0.00s)
=== RUN   TestContainerdParserInvalidType
--- PASS: TestContainerdParserInvalidType (0.00s)
=== RUN   TestFormatDetectionFailure
--- PASS: TestFormatDetectionFailure (0.00s)
=== RUN   TestInternalRecombineCfg
--- PASS: TestInternalRecombineCfg (0.00s)
=== RUN   TestProcess
=== RUN   TestProcess/Success
=== RUN   TestProcess/Success/docker
=== RUN   TestProcess/Success/docker_with_auto_detection
=== RUN   TestProcess/Success/docker_with_auto_detection_and_metadata_from_file_path
=== RUN   TestProcess/Failure
=== RUN   TestProcess/Failure/docker_with_add_metadata_from_filepath_but_not_included
--- PASS: TestProcess (0.00s)
    --- PASS: TestProcess/Success (0.00s)
        --- PASS: TestProcess/Success/docker (0.00s)
        --- PASS: TestProcess/Success/docker_with_auto_detection (0.00s)
        --- PASS: TestProcess/Success/docker_with_auto_detection_and_metadata_from_file_path (0.00s)
    --- PASS: TestProcess/Failure (0.00s)
        --- PASS: TestProcess/Failure/docker_with_add_metadata_from_filepath_but_not_included (0.00s)
=== RUN   TestRecombineProcess
=== RUN   TestRecombineProcess/crio_standalone_with_auto_detection_and_metadata_from_file_path
=== RUN   TestRecombineProcess/crio_standalone_with_auto_detection_and_metadata_from_rotated_file_path
=== RUN   TestRecombineProcess/containerd_standalone_with_auto_detection_and_metadata_from_file_path
=== RUN   TestRecombineProcess/containerd_standalone_with_auto_detection_and_metadata_from_rotated_file_path
=== RUN   TestRecombineProcess/crio_multiple_with_auto_detection_and_metadata_from_file_path
=== RUN   TestRecombineProcess/containerd_multiple_with_auto_detection_and_metadata_from_file_path
=== RUN   TestRecombineProcess/containerd_multiple_with_auto_detection_and_metadata_from_file_path_windows
--- PASS: TestRecombineProcess (0.71s)
    --- PASS: TestRecombineProcess/crio_standalone_with_auto_detection_and_metadata_from_file_path (0.10s)
    --- PASS: TestRecombineProcess/crio_standalone_with_auto_detection_and_metadata_from_rotated_file_path (0.10s)
    --- PASS: TestRecombineProcess/containerd_standalone_with_auto_detection_and_metadata_from_file_path (0.10s)
    --- PASS: TestRecombineProcess/containerd_standalone_with_auto_detection_and_metadata_from_rotated_file_path (0.10s)
    --- PASS: TestRecombineProcess/crio_multiple_with_auto_detection_and_metadata_from_file_path (0.10s)
    --- PASS: TestRecombineProcess/containerd_multiple_with_auto_detection_and_metadata_from_file_path (0.10s)
    --- PASS: TestRecombineProcess/containerd_multiple_with_auto_detection_and_metadata_from_file_path_windows (0.10s)
=== RUN   TestProcessWithDockerTime
=== RUN   TestProcessWithDockerTime/docker
--- PASS: TestProcessWithDockerTime (0.00s)
    --- PASS: TestProcessWithDockerTime/docker (0.00s)
=== RUN   TestCRIRecombineProcessWithFailedDownstreamOperator
=== RUN   TestCRIRecombineProcessWithFailedDownstreamOperator/crio_multiple
=== RUN   TestCRIRecombineProcessWithFailedDownstreamOperator/containerd_multiple
--- PASS: TestCRIRecombineProcessWithFailedDownstreamOperator (0.20s)
    --- PASS: TestCRIRecombineProcessWithFailedDownstreamOperator/crio_multiple (0.10s)
    --- PASS: TestCRIRecombineProcessWithFailedDownstreamOperator/containerd_multiple (0.10s)
=== RUN   TestPathCacheIntegration
=== RUN   TestPathCacheIntegration/CacheEnabled
=== RUN   TestPathCacheIntegration/CacheDisabled
=== RUN   TestPathCacheIntegration/CacheReusesParsedPaths
=== RUN   TestPathCacheIntegration/CacheHandlesDifferentPaths
=== RUN   TestPathCacheIntegration/CacheFIFOEviction
=== RUN   TestPathCacheIntegration/CacheStop
--- PASS: TestPathCacheIntegration (0.00s)
    --- PASS: TestPathCacheIntegration/CacheEnabled (0.00s)
    --- PASS: TestPathCacheIntegration/CacheDisabled (0.00s)
    --- PASS: TestPathCacheIntegration/CacheReusesParsedPaths (0.00s)
    --- PASS: TestPathCacheIntegration/CacheHandlesDifferentPaths (0.00s)
    --- PASS: TestPathCacheIntegration/CacheFIFOEviction (0.00s)
    --- PASS: TestPathCacheIntegration/CacheStop (0.00s)
=== RUN   TestParseCRIOFields
=== RUN   TestParseCRIOFields/valid_crio_with_log
=== RUN   TestParseCRIOFields/valid_crio_stderr
=== RUN   TestParseCRIOFields/valid_crio_no_log_body
=== RUN   TestParseCRIOFields/valid_crio_with_space_after_logtag
=== RUN   TestParseCRIOFields/invalid_missing_stream
=== RUN   TestParseCRIOFields/invalid_invalid_stream
=== RUN   TestParseCRIOFields/invalid_empty
--- PASS: TestParseCRIOFields (0.00s)
    --- PASS: TestParseCRIOFields/valid_crio_with_log (0.00s)
    --- PASS: TestParseCRIOFields/valid_crio_stderr (0.00s)
    --- PASS: TestParseCRIOFields/valid_crio_no_log_body (0.00s)
    --- PASS: TestParseCRIOFields/valid_crio_with_space_after_logtag (0.00s)
    --- PASS: TestParseCRIOFields/invalid_missing_stream (0.00s)
    --- PASS: TestParseCRIOFields/invalid_invalid_stream (0.00s)
    --- PASS: TestParseCRIOFields/invalid_empty (0.00s)
=== RUN   TestParseContainerdFields
=== RUN   TestParseContainerdFields/valid_containerd_with_log
=== RUN   TestParseContainerdFields/valid_containerd_stderr
=== RUN   TestParseContainerdFields/valid_containerd_no_log_body
=== RUN   TestParseContainerdFields/valid_containerd_with_space_after_logtag
=== RUN   TestParseContainerdFields/invalid_no_Z_suffix
=== RUN   TestParseContainerdFields/invalid_missing_stream
=== RUN   TestParseContainerdFields/invalid_invalid_stream
=== RUN   TestParseContainerdFields/invalid_empty
--- PASS: TestParseContainerdFields (0.00s)
    --- PASS: TestParseContainerdFields/valid_containerd_with_log (0.00s)
    --- PASS: TestParseContainerdFields/valid_containerd_stderr (0.00s)
    --- PASS: TestParseContainerdFields/valid_containerd_no_log_body (0.00s)
    --- PASS: TestParseContainerdFields/valid_containerd_with_space_after_logtag (0.00s)
    --- PASS: TestParseContainerdFields/invalid_no_Z_suffix (0.00s)
    --- PASS: TestParseContainerdFields/invalid_missing_stream (0.00s)
    --- PASS: TestParseContainerdFields/invalid_invalid_stream (0.00s)
    --- PASS: TestParseContainerdFields/invalid_empty (0.00s)
=== RUN   TestParseLogPath
=== RUN   TestParseLogPath/valid_unix_path
=== RUN   TestParseLogPath/valid_rotated_log
=== RUN   TestParseLogPath/valid_with_restart_count
=== RUN   TestParseLogPath/valid_uuid_format
=== RUN   TestParseLogPath/invalid_too_short
=== RUN   TestParseLogPath/invalid_missing_pod_part
=== RUN   TestParseLogPath/invalid_not_enough_underscores
=== RUN   TestParseLogPath/invalid_not_log_extension
=== RUN   TestParseLogPath/invalid_non_numeric_restart_count
=== RUN   TestParseLogPath/invalid_empty
--- PASS: TestParseLogPath (0.00s)
    --- PASS: TestParseLogPath/valid_unix_path (0.00s)
    --- PASS: TestParseLogPath/valid_rotated_log (0.00s)
    --- PASS: TestParseLogPath/valid_with_restart_count (0.00s)
    --- PASS: TestParseLogPath/valid_uuid_format (0.00s)
    --- PASS: TestParseLogPath/invalid_too_short (0.00s)
    --- PASS: TestParseLogPath/invalid_missing_pod_part (0.00s)
    --- PASS: TestParseLogPath/invalid_not_enough_underscores (0.00s)
    --- PASS: TestParseLogPath/invalid_not_log_extension (0.00s)
    --- PASS: TestParseLogPath/invalid_non_numeric_restart_count (0.00s)
    --- PASS: TestParseLogPath/invalid_empty (0.00s)
=== RUN   TestSplitOnce
=== RUN   TestSplitOnce/valid_single_separator
=== RUN   TestSplitOnce/valid_multiple_separators
=== RUN   TestSplitOnce/valid_separator_at_start
=== RUN   TestSplitOnce/valid_separator_at_end
=== RUN   TestSplitOnce/invalid_no_separator
=== RUN   TestSplitOnce/invalid_empty
=== RUN   TestSplitOnce/valid_different_separator
--- PASS: TestSplitOnce (0.00s)
    --- PASS: TestSplitOnce/valid_single_separator (0.00s)
    --- PASS: TestSplitOnce/valid_multiple_separators (0.00s)
    --- PASS: TestSplitOnce/valid_separator_at_start (0.00s)
    --- PASS: TestSplitOnce/valid_separator_at_end (0.00s)
    --- PASS: TestSplitOnce/invalid_no_separator (0.00s)
    --- PASS: TestSplitOnce/invalid_empty (0.00s)
    --- PASS: TestSplitOnce/valid_different_separator (0.00s)
PASS
ok  	github.com/open-telemetry/opentelemetry-collector-contrib/pkg/stanza/operator/parser/container	1.240s
```

Nitesh-vaidyanath commented:
Waiting on someone to review changes

Nitesh-vaidyanath commented:
@andrzej-stencel Any luck ?

Nitesh-vaidyanath commented:
Can someone reopen this please ?

atoulme commented on pkg/stanza/operator/helper/cache.go:
```diff
@@ -0,0 +1,261 @@
+)
+
+// Cache allows operators to cache a value and look it up later.
+type Cache interface {
```
this is a lot of new API - the old one was unexported on purpose.

My advice is you move this under `internal` and do not expose this as API externally.

atoulme commented on pkg/stanza/operator/parser/regex/config.go:
just moving this out of the package is worth a PR by itself in preparation of this work - consider splitting this out to a separate PR that can land faster while we discuss applying cache to the container parser.
